Gem sim validation porting 76 x only occupy plots

geonmo opened the pull request:
### Hit Validation module for GEM Detector after Run2
1. This PR include Validation/MuonGEMHits, MuonGEMDigis, MuonGEMRecHits module to validate GEM hit simulation.
   - MuonGEMHits module is related to SimHit on GEM Detector(Run2019/2023).
   - MuonGEMDigis module is related to Digi step about GEM Strip, pad and Copad.
   - MuonGEMRechits module is related to local reco hit on GEM Detector. 
#### This PR does not affect to running other modules.
- These module are only running with specific customize options.

Remake PR due to wrong push to my branch and too old.
It is same PR #10723 except some fix typo or remove commented lines.

jhgoh committed:
fix typo

jhgoh committed:
VID with adjustable cut values

jhgoh committed:
Muon VID are not approved yet

jhgoh committed:
implement value()

geonmo committed:
Add MuonGEM Simhit, digiHit and recoHit occupancy plots

geonmo committed:
Merged gem-sim-validation-porting-76X-only_occupyPlots from repository geonmo

geonmo committed:
Fix some typo when removed efficiency fetecture.

jhgoh committed:
runOnMiniAOD

jhgoh committed:
use patMuons

covarell committed:
revise GEN release validation samples

covarell committed:
Add a missing line

covarell committed:
add powheg ttbar

covarell committed:
add missing file

covarell committed:
add missing file

geonmo committed:
Merged gem-sim-validation-porting-76X-only_occupyPlots from repository geonmo

geonmo committed:
Remove efficiency plots before removing noise.

cmsbuild commented:
A new Pull Request was created by @geonmo (Geonmo Ryu) for CMSSW_7_6_X.

Gem sim validation porting 76 x only occupy plots

It involves the following packages:

Validation/MuonGEMDigis
Validation/MuonGEMHits
Validation/MuonGEMRecHits


@cmsbuild, @danduggan, @deguio can you please review it and eventually sign? Thanks.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

deguio commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 09407a8dd01ca98bb3a60a3c5a3ea75461fed625
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11123/8023/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11123/8023/summary.html

covarell committed:
move some wf to extendedgen so they are not run by the IB

fwyzard committed:
CorrectedJetProducer: change into a global::EDProducer

change all specialisations of `CorrectedJetProducer` into
`global::EDProducer` modules:
  - `CorrectedCaloJetProducer`
  - `CorrectedPFJetProducer`
  - `CorrectedJPTJetProducer`
  - `CorrectedTrackJetProducer`
  - `CorrectedGenJetProducer`

jhgoh committed:
Merged refs/pull/11260/head from repository cms-sw

Commit:
correctedP4 must return a copy of p4()

`correctedJet()` returns a new `Jet` object. In `correctedP4()` we
return a reference to p4() on a stack, thus the reference is invalid and
most likely value will be changed with the next few function calls.

    DataFormats/PatCandidates/interface/Jet.h:155:178: error: function
    returns address of local variable [-Werror=return-local-addr]
    DataFormats/PatCandidates/interface/Jet.h:158:204: error: function
    returns address of local variable [-Werror=return-local-addr]

A quick fix (this patch) is to make `correctedP4()` return a copy
instead of a reference.

If `correctedJet()` is on hot path, the new `Jet` object should be
cached and then we could also return a reference in `correctedP4()`.

Signed-off-by: David Abdurachmanov <David.Abdurachmanov@cern.ch>

jhgoh committed:
fix typo

jhgoh committed:
do not use boost

jhgoh committed:
use additional inputTag for miniAOD

jhgoh committed:
split dxy and dz cuts

jhgoh committed:
udpate MD5

jhgoh committed:
make one line

fwyzard committed:
replace GNU/C99 VLA with std::vector

jhgoh committed:
protect from division by zero, avoid confusion bool <-> double

jhgoh committed:
avoid confusion in return type bool <-> double

geonmo commented:
@deguio Hello, Federico. Could you approve this PR? I will make another PR for efficiency plots.

davidlange6 committed:
Merge pull request #11323 from jhgoh/MuonVID76X

Muon VID update for 76X

cmsbuild committed:
Merge pull request #11396 from hengne/from-CMSSW_7_6_0_pre5_relval_rebase

relval scripts update

cmsbuild committed:
Merge pull request #11311 from davidlt/fix-Jet-return-local-addr

correctedP4 must return a copy of p4()

cmsbuild committed:
Merge pull request #11004 from covarell/76X_reviseGenVal

revise GEN release validation samples

cmsbuild committed:
Merge pull request #11301 from fwyzard/global_CorrectedJetProducer_76x

CorrectedJetProducer: change into a global::EDProducer (76x)

rovere commented on Validation/MuonGEMDigis/src/GEMCheckGeometry.cc:
```diff
@@ -0,0 +1,105 @@
+        temp_me->setBinLabel(10,"St3,La1_odd",2);
+        temp_me->setBinLabel(11,"St3,La2_even",2);
+        temp_me->setBinLabel(12,"St3,La2_odd",2);
+        theStdPlots.insert( std::map< UInt_t, MonitorElement*>::value_type(name.Hash(), temp_me ));
```
I do not see the point of using an ordinary map as a hash-table?? What is the meaning of having the hash-keys ordered (and the time spent to order them)? You get the worst of the 2 from this approach, IMHO. Use directly an unordered_map.

geonmo replied:
Hello, rovere. I used "hash" function to get unique value from "string". Frankly speaking, I have a some trouble due to "const char*" comparision problem. So, I used Hash function to solve temporary. In addition, I heared unorderd_map is slower than map for small pairs. I will fix to use "map" with "std::string". Thanks for interest.

rovere replied:
No, using std:string as key in the map is even worse. Either leave it this way, or use an std::unordered_map with a string as key: string should natively have hashing specialisation available to you. And std hash tables handle collisions for you.

deguio commented:
ciao @geonmo 
I am in the process of reviewing this PR finally..  (sorry for being so late)
a couple of general comments in addition to marco's comment:

- there are few cases where memory is allocated with new but never deleted. see for example:
https://github.com/cms-sw/cmssw/pull/11123/files#diff-0b49d714d6e00574dcfa3e3e57202922R63
- I am not sure which criterion you used to name cfi vs cff files
- the number of histograms (bins) booked is reasonable. I hope that the detailed plots will be considered for local running only. see for example: https://github.com/cms-sw/cmssw/pull/11123/files#diff-b0e32bae891ed958cf3bf1132d1b22c6R33  (4.320.000 bins!)
- is this class under development? huge portion of code is commented out. https://github.com/cms-sw/cmssw/pull/11123/files#diff-5e6117ecb55ff174214e1cabd9f2cb3fR97


This PR adds code which is not included in any central sequence, but has been tested locally. more details at:
https://indico.cern.ch/event/446270/contribution/2/attachments/1154267/1658925/Presentation_on_DQM_MeetingGEM_Validation_modules.pdf

geonmo commented:
@deguio Hello, Federico.  
- For Comment #1, 
 - Ok , I will fix it as soon as possible. However, it is a big holyday season in Korea. So, I can update next weeks or later.
- For Comment #2, 
 - I used cfi files for dummy values(to show all variables) and cff files for real values(to hide some variables). However, they were written discursively. I will cleaning and refactoring to proper standard.
- For Comment #3,
 - detailPlots are only used for gem software developer. It never run on central process.(If so, it is a bug.)
- For Comment #4, 
 - That file is related to efficiency plots(actullay, it is used to draw detail efficiency plots.). I commented out for this PR. However, I will uncomment and fix when I make a PR for efficiency plots.

Many thanks for interest and review.
I will update as soon as possible as your comment. 
Thanks,

geonmo commented:
Hello, @deguio 
I checked about to do not delete memory allocation.
I made some histogram and profile using "new" keyword and all cases will be booked on IBooker.
As I know, ibooker's book functions need histogram's pointers and ibooker will deleted if no more histogram is required, isn't it?

geonmo committed:
Merged gem-sim-validation-porting-76X-only_occupyPlots from repository geonmo

geonmo committed:
Refactoring and bugfix as comments.

geonmo commented:
Hello, @deguio and @rovere .
I took a mistake to missing rebase before submitting a PR.
I will re-make new PR with same contents.
